TFMarian, TFMbart, TFPegasus, TFBlenderbot

sshleifer opened the pull request:
### Notes:
- add `TFSinusoidalPositionalEmbeddings`.
- Code structure identical to the corresponding pytorch code -- same classes, implementations differ only slightly.
- Integration tests and common tests, rst updates, for all 4 children. All 4 children run the same common tests as TFBart and at least 1 integration test.
- For pegasus, generations are not identical to PT because Linear layers are slightly different in tf/pt. For Marian, generations are identical.
- Loading will generate 0 warnings.

LysandreJik commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -262,7 +261,7 @@ def call(self, x, encoder_padding_mask, training=False):
         x = self.activation_fn(self.fc1(x))
         x = self.activation_dropout(x, training=training)
         x = self.fc2(x)
-        x = self.dropout_wt(x, training=training)
+        x = tf.nn.dropout(x, rate=self.dropout if training else 0)
```
You're defining two ways of doing the dropout here, can this be consistent?
Shouldn't this just be
```py
x = self.dropout(x, training=training)
```
since you initialized `dropout` on line 229? (i.e., exactly as it was before?)

sshleifer replied:
yes, will fix.

sshleifer replied:
switched to always using `tf.nn.dropout(x, rate=x if training else 0)`

LysandreJik commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -458,16 +467,19 @@ def call(
+        if self.normalize_before:
+            x = self.final_layer_norm(x)
         x = self.activation_fn(self.fc1(x))
         x = tf.nn.dropout(x, rate=self.activation_dropout if training else 0)
         x = self.fc2(x)
         x = tf.nn.dropout(x, rate=self.dropout if training else 0)
```
Had not seen last review, why use `tf.nn.dropout`  and `tf.keras.layers.Dropout` and not stick with a single one?

sshleifer replied:
The reason was pytorch consistency, but I will fix both.

LysandreJik commented on src/transformers/modeling_tf_blenderbot.py:
```diff
@@ -0,0 +1,41 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TF BART model, ported from the fairseq repo."""
```
```suggestion
"""TF BlenderBot model, ported from the fairseq repo."""
```

LysandreJik commented on src/transformers/modeling_tf_blenderbot.py:
```diff
@@ -0,0 +1,41 @@
+logger = logging.get_logger(__name__)
+
+
+@add_start_docstrings("Marian model for machine translation", START_DOCSTRING)
```
Don't think that's the correct docstring here

LysandreJik commented on src/transformers/modeling_tf_marian.py:
```diff
@@ -0,0 +1,48 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TF BART model, ported from the fairseq repo."""
```
Marian
```suggestion
"""TF Marian model, ported from the fairseq repo."""
```

LysandreJik commented on src/transformers/modeling_tf_pegasus.py:
```diff
@@ -0,0 +1,41 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TF BART model, ported from the fairseq repo."""
```
```suggestion
"""TF Pegasus model, ported from the fairseq repo."""
```

LysandreJik commented on utils/check_repo.py:
```diff
@@ -67,6 +67,7 @@
     "xlm_prophetnet": "xlmprophetnet.rst",
     "xlm_roberta": "xlmroberta.rst",
     "bert_generation": "bertgeneration.rst",
+    "marian": "marian.rst",
```
Is this necessary?

sgugger replied:
Shoudn't be

sshleifer replied:
I get 
```python
Exception: There were 1 failures:
transformers.modeling_tf_marian does not have its corresponding doc file marian.rst. If the doc file exists but isn't named marian.rst, update `MODEL_NAME_TO_DOC_FILE` in the file `utils/check_repo.py`.
```
without. Will investigate.

sgugger replied:
Don't spend too much time if it's my script giving you a hard time. I can investigate after this PR is merged.

sgugger reviewed (approved):
Thanks for adding those models. LGTM modulo @LysandreJik doc comments

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -289,26 +288,30 @@ def __init__(self, config: BartConfig, embed_tokens: TFSharedEmbeddings, **kwarg
-            config.extra_pos_embeddings,
-            name="embed_positions",
-        )
+        if config.static_position_embeddings:
```
did we have the name `static_position_embeddings` already before? Think a better name would be `use_sinusoidal_pos_embeddings`

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -289,26 +288,30 @@ def __init__(self, config: BartConfig, embed_tokens: TFSharedEmbeddings, **kwarg
-            if config.add_final_layer_norm
-            else None
+        self.layernorm_embedding = (
+            LayerNormalization(epsilon=1e-5, name="layernorm_embedding") if config.normalize_embedding else Layer()
```
This looks a bit hacky here - why do we set `self.layernorm_embedding = Layer()` in `config.normalize_embedding` is `False`?

Guess that's model specific then?

patrickvonplaten replied:
Why not set to `None` as is done below for `self.layer_norm`?

sshleifer replied:
It copies torch. I like moving the conditionals to `__init__` cause it saves a line, saves an if, and only gets evaluated once.

it's the say as 
```python
if self.layer_norm  is not None:
    x = self.layer_norm(x)
```

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -433,10 +437,12 @@ def call(
 
             Tuple containing, encoded output of shape `(seq_len, batch, embed_dim)`, self_attn_weights, layer_state
         """
+        residual = x  # Make a copy of the input tensor to add later.
```
don't like `x` => should be `hidden_states` IMO, but I guess we had this discussion often enough :D

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -780,6 +793,57 @@ def call(self, input_ids: tf.Tensor, use_cache=False):
         return super().call(positions + self.offset)  # super object is not callable for some reason
 
 
+class TFSinusoidalPositionalEmbedding(TFSharedEmbeddings):
```
Why are the positional embeddings `TFSharedEmbeddings`? These weights don't seem to be shared if I read the code correctly, *i.e.* both the encoder and decoder instantiate position ids, but as far as I can tell they don't share their weights. I don't think we need the `TFSharedEmbeddings` abstraction here - or maybe there is  something I'm not seeing?

sshleifer replied:
great catch will use `TFEmbeddings`.

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -1130,15 +1199,17 @@ def adjust_logits_during_generation(self, logits, cur_len, max_length):
         output_list = []
-
         # Is there a better way to do scores[:, [x for if x != token_id]] = -float("inf") in TF?
         bs, vocab_size = scores.shape
```
We should not use `.shape` in TF, but the `shape_list()` function to not run into problems for compilation.

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -252,7 +251,7 @@ def call(self, x, encoder_padding_mask, training=False):
             x = self.self_attn_layer_norm(x)
         x, self_attn_weights = self.self_attn(query=x, key=x, key_padding_mask=encoder_padding_mask)
         assert x.shape == residual.shape, f"Self attn modified the shape of query {residual.shape} to {x.shape}"
```
better use `shape_list()` here

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -780,6 +793,57 @@ def call(self, input_ids: tf.Tensor, use_cache=False):
+
+    def call(self, input_ids, use_cache=False):
+        """Input is expected to be of size [bsz x seqlen]."""
+        bsz, seq_len = input_ids.shape[:2]
```
better to use `shape_list()`

sshleifer replied:
@stas00 the magician do you know how to replace every occurrence of 
`X.shape` with `shape_list(X)` for arbitrary X?

stas00 replied:
sure, but I will need some samples to know how to restrict the search. 

If for example X doesn't contain spaces, then it's easy:
```
perl -pi -e 's|(\S+)\.shape|shape_list($1)|' files
```

sshleifer replied:
Hugely helpful, thanks. It went 23/25. Only issues were multiple in same line.
x is a python variable. @patrickvonplaten  this may be useful to you in future!

stas00 replied:
oh, just add `|g` for global:
```
perl -pi -e 's|(\S+)\.shape|shape_list($1)|g' files
```
glad to hear it worked.

patrickvonplaten commented on src/transformers/modeling_tf_bart.py:
```diff
@@ -1130,15 +1199,17 @@ def adjust_logits_during_generation(self, logits, cur_len, max_length):
 
     @staticmethod
-    def _force_token_id_to_be_generated(scores, token_id) -> None:
+    def _force_token_id_to_be_generated(scores, token_id, inverted=False) -> None:
```
why do we need the flag `inverted` here?

patrickvonplaten replied:
I see `Blenderbot` probably... this function is really not the easiest to understand and tweaking it to make it use for other models is not the way to go here IMO...

patrickvonplaten commented on src/transformers/modeling_tf_marian.py:
```diff
@@ -0,0 +1,48 @@
+
+    def adjust_logits_during_generation(self, logits, cur_len, max_length):
+        """Never predict pad_token_id. Predict </s> when max_length is reached."""
+        self._force_token_id_to_be_generated(logits, self.config.pad_token_id, inverted=True)
```
same here I would rewrite the `force_token_id_to_be_generated` function. `Generate()` is super hard to follow if the code is constantly jumping around files

sshleifer replied:
OK

sshleifer replied:
In parallel, I'm trying to find a better solution to make the code better, https://stackoverflow.com/questions/64575346/tensorflow-set-column-of-tensor-to-infinity

sshleifer replied:
Thanks to messieur Plu I was able to delete `_force_token_id_to_be_generated`.
Now there will be less jumping.

patrickvonplaten commented on tests/test_modeling_tf_bart.py:
```diff
@@ -28,12 +30,16 @@
 
 @require_tf
-class ModelTester:
+class TFBartModelTester:
```
yaaay :-)

sshleifer replied:
on this PR you have to say TFBartyaay or people won't understand what's going on.

patrickvonplaten commented on tests/test_modeling_tf_blenderbot.py:
```diff
@@ -0,0 +1,132 @@
+        pass
+
+    def test_saved_model_with_hidden_states_output(self):
+        # Should be uncommented during patrick TF refactor
```
I give my best haha

patrickvonplaten reviewed (commented):
Love the tests + awesome that 4 new models are added in TF this way!!!

Things I think we could approve upon:

I'm not a fan of how the code is entangled between all the models (*e.g.* adding the `config.normalize_before` statements), but I guess since it's the same in PT it's ok for now!

I would advocate for rewriting the  `_force_token_id_to_be_generated` for the respective models however. The function is really not that long and I don't think we should add a hacky flag to the function to make it work with other models. Plus the `adjust_logits_during_generation` is already a hacky function and now for models like TFMarian the user jumps around a lot between different files.

The other thing I'm curious about is why we have to abstract the position embedding weights from `TFSharedEmbedding` - I don't think this is necessary as written above.

sshleifer commented:
@patrickvonplaten 
+ I deleted the `_force_token_id` function, replaced with faster `tf.where` one-liner. (+ added regression test). 
+ Replaced unneeded `TFSharedEmbedding` with `tf.keras.layers.Embedding`
+ switched all `.shape` to `shape_list`

WDYT?

patrickvonplaten reviewed (approved):
Looks great! Thanks a lot for adding the models and applying some changes :-)

LysandreJik reviewed (approved):
Thanks for applying the changes @sshleifer! LGTM!
